### PR TITLE
Added override variable and additional srvgrp

### PR DESCRIPTION
**Pull request by Xartos**
Draft fix to add server groups per node and local disk per node. Is this a feasible solution? I'll need to add some more for the partition things, but just for the override parts.

Note that I needed to add a experimental feature to make this work

### CODE REVIEW

**Comment by Xartos** [comment]
Okey I think I'm done with the changes now. I'll open the PR here now, but I will not merge here. I'll cherry-pick the changes to 2.20 so we can create a new patch release for that. But I'll wait for this to be approved first

**Comment by aarnq** [comment]
Really nice with the additional server groups, though I don't think I know enough about these stuff to review it properly.
I tried to just run plan against my current cluster (Elastx) and it errors due to "unsupported argument" for `bastion_allowed_ports`, `network.subnet_id` and your `additional_server_groups`.

**Comment by Xartos** [comment]
> Really nice with the additional server groups, though I don't think I know enough about these stuff to review it properly. I tried to just run plan against my current cluster (Elastx) and it errors due to "unsupported argument" for `bastion_allowed_ports`, `network.subnet_id` and your `additional_server_groups`.

Hmm that's strange, I havent seen that I think. But since this is based on master, could you test using this branch instead? https://github.com/elastisys/kubespray/tree/fli/0.20-add-override-and-split-srvgrp

It's based on 2.20.0

**Comment by aarnq** [comment]
> Hmm that's strange, I havent seen that I think. But since this is based on master, could you test using this branch instead? https://github.com/elastisys/kubespray/tree/fli/0.20-add-override-and-split-srvgrp

My bad, apparently `terraform init` wasn't enough so once I cleared everything out then it worked.
So it looks good to me.

**Review by OlleLarsson — APPROVED** [review]
LGTM 🎉  !

Testing done from my side on elastx:
- Vanilla - aka same config as we use now in our environments -> no diff 
- Add additional server groups on specific nodes
- Test partition feature
```
k8s_nodes = {
    "db-3" = {
    "az"          = "sto3"
    "flavor"      = "6b576115-1e0d-4f8e-bc3a-8d45732a44b5" #2C-4GB-60GB
    "floating_ip" = true
    "cloudinit"   = {
      "extra_partitions" = [{
        "volume_path"     = "/dev/vda"
        "partition_path"  = "/dev/vda2"
        "partition_start" = "20GB"
        "partition_end"   = "-1"
        "mount_path"      = "/mnt/local-disk"
      }]
    }
    "server_group" = "test"
  },
    "db-4" = {
    "az"                    = "sto3"
    "flavor"                = "6b576115-1e0d-4f8e-bc3a-8d45732a44b5" #2C-4GB-60GB
    "floating_ip"           = true
    "volume_type"           = "v1-dynamic-40"
    "root_volume_sie_in_gb" = 50
    "server_group"          = "test"
  }
}

additional_server_groups = {
  test = {
    policy = "anti-affinity"
  }
}
```

**Review by aarnq — APPROVED** [review]

File: contrib/terraform/openstack/README.md [hunk 1]
```diff
@@ -284,6 +284,7 @@ For your cluster, edit `inventory/$CLUSTER/cluster.tfvars`.
 |`master_server_group_policy` | Enable and use openstack nova servergroups for masters with set policy, default: "" (disabled) |
 |`node_server_group_policy` | Enable and use openstack nova servergroups for nodes with set policy, default: "" (disabled) |
 |`etcd_server_group_policy` | Enable and use openstack nova servergroups for etcd with set policy, default: "" (disabled) |
+|`additional_server_groups` | Extra server groups to create. Set "policy" to the policy for the group, default: [] (to not create any extra groups) |
```
Comment (viktor-f) [inline contrib/terraform/openstack/README.md:287]: Could you specify the exact format here? Something like this (but with the actually correct format):
```suggestion
|`additional_server_groups` | Extra server groups to create. Set "policy" to the policy for the group, expected format is [{"new-server-group" = {"policy" = "hard-anti-affinity"}}], default: [] (to not create any extra groups) |
```

**Review by viktor-f — APPROVED** [review]

**Review by robinAwallace — APPROVED** [review]
I have not tested it, but it looks good :+1:

**Comment by Xartos** [comment]
Separate PR for the a new patch 2.20.0-ck8s3 is already merged https://github.com/elastisys/kubespray/pull/5
So I'll close this one in favor of https://github.com/kubernetes-sigs/kubespray/pull/9452

**Pull request closed without merging**